src/opensynth/models/faraday/gaussian_mixture/gmm_metrics.py

```diff
@@ -0,0 +1,99 @@
+# Copyright Contributors to the Opensynth-energy Project.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+from torchmetrics import Metric
+
+
+class WeightsMetric(Metric):
+    full_state_update = False
+
+    def __init__(self, num_components: int):
+        super().__init__()
+        self.weights: torch.Tensor
+
+        self.add_state(
+            "weights", torch.zeros(num_components), dist_reduce_fx="mean"
+        )
+
+    def update(self, weights: torch.Tensor) -> None:
+        self.weights.add_(weights)
+
+    def compute(self) -> None:
+        return self.weights
+
+
+class MeansMetric(Metric):
+    full_state_update = False
+
+    def __init__(self, num_components: int, num_features: int):
+        super().__init__()
+        self.means: torch.Tensor
+        self.add_state(
+            "means",
+            torch.zeros(num_components, num_features),
+            dist_reduce_fx="mean",
+        )
+
+    def update(self, means: torch.Tensor) -> None:
+        self.means.add_(means)
+
+    def compute(self) -> None:
+        return self.means
+
+
+class PrecisionCholeskyMetric(Metric):
+    full_state_update = False
+
+    def __init__(self, num_components: int, num_features: int):
+        super().__init__()
+        self.precision_cholesky: torch.Tensor
+        self.add_state(
+            "precision_cholesky",
+            torch.zeros(num_components, num_features, num_features),
+            dist_reduce_fx="mean",
+        )
+
+    def update(self, precision_cholesky: torch.Tensor) -> None:
+        self.precision_cholesky.add_(precision_cholesky)
+
+    def compute(self) -> None:
+        return self.precision_cholesky
+
+
+class CovarianceMetric(Metric):
+    full_state_update = False
+
+    def __init__(self, num_components: int, num_features: int):
+        super().__init__()
+        self.covariances: torch.Tensor
+        self.add_state(
+            "covariances",
+            torch.zeros(num_components, num_features, num_features),
+            dist_reduce_fx="mean",
+        )
+
+    def update(self, covariances: torch.Tensor) -> None:
+        self.covariances.add_(covariances)
+
+    def compute(self) -> None:
+        return self.covariances
+
+
+class NLLMetric(Metric):
```

Comment:
I think maybe we should called this `LogProbMetric` instead, as this is strictly speaking the log prob, rather than NLL. Unless you want to move the NLL calculation here into this metric, then we call it `NLLMetric`. 

Also I prefer we call it `NegativeLogLikelihoodMetric` too to and avoid abbreviations to make the code a lot more self-descriptive.

Reply:
Just read further down the line that the value we're passing into this metric is already the negative log likelihood. Would change the attribute `self.log_prob` to `self.nll` instead to make this clearer as the class name doesn't match its attributes.

Reply:
Updated a suggested.